File: dotnet/OrisunIbukun/Utils/Validators.cs
using System;

namespace OrisunIbukun.Utils
{
    public static class Validators
    {
        public static (bool valid, double amount) ValidateAmount(string value)
        {
            if (string.IsNullOrWhiteSpace(value))
                return (false, 0);

            string cleaned = value.Replace("₦", "").Replace(",", "").Trim();
            if (double.TryParse(cleaned, out double amount) && amount > 0)
                return (true, amount);

            return (false, 0);
        }

        public static (bool valid, string error) ValidateRequired(string value, string fieldName)
        {
            if (string.IsNullOrWhiteSpace(value))
                return (false, $"{fieldName} is required");
            return (true, "");
        }

        public static (bool valid, string error) ValidateDate(string value, string fieldName)
        {
            if (string.IsNullOrWhiteSpace(value))
                return (false, $"{fieldName} is required");
            if (!DateTime.TryParse(value, out _))
                return (false, $"{fieldName} must be a valid date");
            return (true, "");
        }

        public static (bool valid, string error) ValidatePin(string value)
        {
            if (string.IsNullOrWhiteSpace(value))
                return (false, "PIN is required");
            string pin = value.Trim();
            if (pin.Length < 4 || pin.Length > 6 || !IsAllDigits(pin))
                return (false, "PIN must be 4-6 digits");
            return (true, "");
        }

        public static (bool valid, string error) ValidatePhone(string value, string fieldName = "Phone")
        {
            if (string.IsNullOrWhiteSpace(value))
                return (true, "");
            int digits = 0;
            foreach (char c in value.Trim())
                if (char.IsDigit(c)) digits++;
            if (digits < 7)
                return (false, $"{fieldName} must contain at least 7 digits");
            return (true, "");
        }

        public static (bool valid, string error) ValidateName(string value, string fieldName = "Full Name")
        {
            if (string.IsNullOrWhiteSpace(value))
                return (false, $"{fieldName} is required");
            string name = value.Trim();
            if (name.Length < 2)
                return (false, $"{fieldName} must be at least 2 characters");
            foreach (char c in name)
            {
                if (!(char.IsLetter(c) || char.IsDigit(c) || c == ' ' || c == '.' || c == '-' || c == '\'' || c == ','))
                    return (false, $"{fieldName} contains invalid characters");
            }
            return (true, "");
        }

        public static (bool valid, string error) ValidateEmail(string value)
        {
            if (string.IsNullOrWhiteSpace(value))
                return (true, "");
            string email = value.Trim().ToLower();
            int at = email.IndexOf('@');
            if (at <= 0 || at != email.LastIndexOf('@') || !email.Substring(at).Contains("."))
                return (false, "Invalid email address");
            return (true, "");
        }

        private static bool IsAllDigits(string value)
        {
            if (value.Length == 0) return false;
            foreach (char c in value)
                if (!char.IsDigit(c)) return false;
            return true;
        }
    }
}

File: app/utils/validators.py
"""Validation helpers."""
import re
import html


def sanitize_string(value: str, max_length: int | None = None) -> str:
    """Sanitize string input - strip whitespace and escape HTML."""
    value = value.strip()
    value = html.escape(value)
    if max_length and len(value) > max_length:
        value = value[:max_length]
    return value


def validate_member_id(member_id: str) -> bool:
    return bool(re.match(r"^ORI-\d{5}$", member_id))


def validate_amount(value: str) -> float:
    cleaned = value.replace(",", "").replace("₦", "").strip()
    if not cleaned:
        raise ValueError("Amount is required.")
    try:
        amount = float(cleaned)
    except ValueError:
        raise ValueError("Please enter a valid number.")
    if amount < 0:
        raise ValueError("Amount cannot be negative.")
    return amount


def validate_positive_amount(value: str) -> float:
    """Validate amount is positive (greater than zero)."""
    amount = validate_amount(value)
    if amount <= 0:
        raise ValueError("Amount must be greater than zero.")
    return amount


def validate_phone(phone: str) -> bool:
    cleaned = phone.strip().replace(" ", "").replace("-", "")
    return len(cleaned) >= 7 and cleaned.isdigit()


def validate_phone_strict(phone: str) -> str:
    """Validate and return cleaned phone number."""
    cleaned = phone.strip().replace(" ", "").replace("-", "").replace("(", "").replace(")", "")
    if not cleaned:
        raise ValueError("Phone number is required.")
    if not cleaned.isdigit():
        raise ValueError("Phone number must contain only digits.")
    if len(cleaned) < 7:
        raise ValueError("Phone number must be at least 7 digits.")
    return cleaned


def validate_email(email: str) -> str:
    """Validate email format."""
    email = email.strip().lower()
    if not email:
        raise ValueError("Email is required.")
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if not re.match(pattern, email):
        raise ValueError("Please enter a valid email address.")
    return email


def validate_required(value: str, field_name: str) -> str:
    value = value.strip()
    if not value:
        raise ValueError(f"{field_name} is required.")
    return value


def validate_pin(pin: str) -> str:
    """Validate PIN - must be 4-6 digits."""
    pin = pin.strip()
    if not pin:
        raise ValueError("PIN is required.")
    if not pin.isdigit():
        raise ValueError("PIN must contain only digits.")
    if len(pin) < 4 or len(pin) > 6:
        raise ValueError("PIN must be 4-6 digits.")
    return pin


def validate_name(name: str, field_name: str = "Name") -> str:
    """Validate name - Unicode letters (incl. diacritics), digits, spaces, hyphens, apostrophes, periods."""
    name = name.strip()
    if not name:
        raise ValueError(f"{field_name} is required.")
    if len(name) < 2:
        raise ValueError(f"{field_name} must be at least 2 characters.")
    for ch in name:
        if not (ch.isalpha() or ch.isdigit() or ch in " .-'"):
            raise ValueError(f"{field_name} contains an invalid character: {ch!r}.")
    return name


def sanitize_input(value: str, max_length: int = 500) -> str:
    """General purpose input sanitization."""
    return sanitize_string(value, max_length)
